src/UrlSearchParams.ts

```diff
@@ -0,0 +1,151 @@
+import { duplicate } from './lang';
+
+/**
+ * A object with string keys and string or string array values that describes a query string.
+ */
+type ParamList = { [key: string]: string | string[] };
+
+/**
+ * Parse a query string, returning a ParamList object.
+ */
+function parseQueryString(input: string): ParamList {
+	const query = <{ [key: string]: string[] }> {};
+	input.split('&').forEach(entry => {
+		let [ key, value ] = entry.split('=');
+		key = key ? decodeURIComponent(key) : '';
+		value = value ? decodeURIComponent(value) : '';
+
+		if (key in query) {
+			query[key].push(value);
+		}
+		else {
+			query[key] = [ value ];
+		}
+	});
+	return query;
+}
+
+/**
+ * An class that represents a URL query search parameters.
```

Comment:
Apparently grammar had just totally abandoned me by the time I got to block comments.